src/System.Text.CaseFolding/System/Text/CaseFolding/SimpleCaseFolding.cs

```diff
@@ -0,0 +1,630 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Text.CaseFolding
+{
+    /// <summary>
+    /// Implement Unicode Simple Case Folding API.
+    /// </summary>
+    public static partial class SimpleCaseFolding
+    {
+        private static ref ushort s_MapLevel1 => ref MapLevel1[0];
+        private static ref char s_refMapData => ref MapData[0];
+        private static ref ushort s_refMapSurrogateLevel1 => ref MapSurrogateLevel1[0];
+        private static ref (char highSurrogate, char lowSurrogate) s_refMapSurrogateData => ref MapSurrogateData[0];
+
+        /// <summary>
+        /// Simple case fold the char.
+        /// </summary>
+        /// <param name="c">Source char.</param>
+        /// <returns>
+        /// Returns folded char.
+        /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static char SimpleCaseFold(char c)
+        {
+            if (c <= 0x5ff)
+            {
+                return MapBelow5FF[c];
+            }
+
+            // var v = L1[c >> 8];
+            // var ch = L3[v + (c & 0xFF)];
+            // Still slow due to border checks.
+            var v = Unsafe.Add(ref s_MapLevel1, c >> 8);
+            var ch = Unsafe.Add(ref s_refMapData, v + (c & 0xFF));
+
+            return ch == 0 ? c : ch;
+        }
+
+        // Mapping for chars > 0x5ff slowly due to 2-level mapping.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int SimpleCaseFoldCompareAbove05ff(char c1, char c2, ref ushort refMapLevel1, ref char refMapData)
+        {
+            Debug.Assert((int)c1 > 0x5ff, "Char should be greater than 0x5ff.");
+            Debug.Assert((int)c2 > 0x5ff, "Char should be greater than 0x5ff.");
+            var v1 =  Unsafe.Add(ref refMapLevel1, c1 >> 8);
+            var ch1 = Unsafe.Add(ref refMapData, v1 + (c1 & 0xFF));
```

Comment:
As I said above, this code is very sensitive to changes. I would leave this to MSFT experts to see what is happening there in-depth.

Reply:
Moving to helper method that is marked as AggressivelyInlined shouldn't noticeably affect performance and will reduce **a lot** of code duplication.

Reply:
TC doesn't inline at tier0 so I expect the change decrease performance.

Reply:
We should measure perf with tiered jitting disabled. If the method is on a hot path, it would be promoted to a higher tier anyway. If it isn't on a hot path, the perf optimization is less critical.

You can disable tiered jitting in your benchmark by setting a custom environment variable (`COMPlus_TieredCompilation` = 0).

For BenchmarkDotNet, see this custom config as an example:
https://github.com/dotnet/corefxlab/blob/76cd6c2f7739798605d8f6bd6566581bce2ee6a9/tests/Benchmarks/System.Text.JsonLab/JsonWriterPerf.cs#L18-L33

Reply:
Thanks for the pattern. Can we postpone to follow PR as I asked above?